org.alloytools.alloy.classic.provider/src/main/java/org/alloytools/alloy/classic/provider/Relation.java

```diff
@@ -0,0 +1,194 @@
+package org.alloytools.alloy.classic.provider;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import org.alloytools.alloy.core.api.IAtom;
+import org.alloytools.alloy.core.api.IRelation;
+import org.alloytools.alloy.core.api.ITuple;
+import org.alloytools.alloy.core.api.Solution;
+
+public class Relation implements IRelation {
+
+    final Solution solution;
+    final int      arity;
+    final ITuple[] tuples;
+
+    Relation(Solution solution, int arity, Tuple[] tuples) {
+        this.solution = solution;
+        this.tuples = tuples;
+        this.arity = arity;
+    }
+
+    public Relation(Solution solution, int arity, List<? extends IAtom> atoms) {
+        this(solution, arity, toTuples(solution, arity, atoms));
+    }
+
+    @Override
+    public int arity() {
+        return arity;
+    }
+
+    @Override
+    public IRelation join(IRelation right) {
+        assert solution == right.getSolution();
+
+        int arity = this.arity() + right.arity() - 2;
+        List<IAtom> atoms = new ArrayList<>();
+
+        for (ITuple l : this) {
+            IAtom last = l.last();
+
+            for (ITuple r : right) {
+                IAtom first = r.first();
+
+                if (last == first) {
+                    for (int i = 0; i < l.arity() - 1; i++) {
+                        atoms.add(l.get(i));
+                    }
+
+                    for (int i = 1; i < r.arity(); i++) {
+                        atoms.add(r.get(i));
+                    }
+                }
+            }
+        }
+
+        return new Relation(solution, arity, atoms);
+    }
+
+    @Override
+    public IRelation product(IRelation right) {
+        assert solution == right.getSolution();
+
+        List<IAtom> atoms = new ArrayList<>();
+        int arity = this.arity() + right.arity();
+        for (ITuple l : this) {
+            for (ITuple r : right) {
+                for (int i = 0; i < l.arity(); i++) {
+                    atoms.add(l.get(i));
+                }
+                for (int i = 0; i < r.arity(); i++) {
+                    atoms.add(r.get(i));
+                }
+            }
+        }
+        return new Relation(solution, arity, atoms);
+    }
+
+    @Override
+    public Relation head() {
+        return split(0, 1);
```

Comment:
the `split` method says that `from` must be greater than 0 and here you're passing `0`